### PR TITLE
[CELEBORN-1636][FOLLOWUP] Dynamic resources will only be utilized in case of candidates shortages

### PR DESCRIPTION
**Pull request by zaynt4606**
### What changes were proposed in this pull request?
Follow up of [https://github.com/apache/celeborn/pull/2835]
Only use dynamic resources when candidates are not enough.
And change the way geting availableWorkers form heartbeat to requestSlots RPC to avoid the burden of heartbeat.

### Why are the changes needed?
No

### Does this PR introduce _any_ user-facing change?
Add another configuration.

### How was this patch tested?
UT

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -4838,6 +4839,17 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(false)
 
+  val CLIENT_SHUFFLE_DYNAMIC_RESOURCE_FACTOR: ConfigEntry[Double] =
+    buildConf("celeborn.client.shuffle.dynamicResourceFactor")
+      .categories("client")
+      .version("0.6.0")
+      .doc("The ChangePartitionManager will check whether (unavailable workers / shuffle allocated workers) " +
+        "is more than the factor before obtaining candidate workers from the availableWorkers pool “ +" +
```
Comment (turboFei) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4847]: `pool “ +" ` seems typo

Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Thanks! Fixed~

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -4838,6 +4839,17 @@ object CelebornConf extends Logging {
...
+  val CLIENT_SHUFFLE_DYNAMIC_RESOURCE_FACTOR: ConfigEntry[Double] =
+    buildConf("celeborn.client.shuffle.dynamicResourceFactor")
+      .categories("client")
+      .version("0.6.0")
+      .doc("The ChangePartitionManager will check whether (unavailable workers / shuffle allocated workers) " +
+        "is more than the factor before obtaining candidate workers from the availableWorkers pool " +
+        "during heartbeats when worker resources change and CLIENT_SHUFFLE_DYNAMIC_RESOURCE_ENABLED set true")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v <= 1.0, "Should be in [0.0, 1.0].")
+      .createWithDefault(0.25)
```
Comment (turboFei) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4851]:
```
and ${CLIENT_SHUFFLE_DYNAMIC_RESOURCE_ENABLED.key} set to true
```

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 3]
```diff
@@ -291,42 +292,71 @@ class ChangePartitionManager(
...
+        .asJava
+    candidates.addAll(snapshotCandidates)
+
+    if (clientShuffleDynamicResourceEnabled) {
+      val shuffleAllocatedWorkers = lifecycleManager.workerSnapshots(shuffleId).size()
+      val unavailableWorkerRatio = 1 - (snapshotCandidates.size * 1.0 / shuffleAllocatedWorkers)
+      if (candidates.size < 1 || (pushReplicateEnabled && candidates.size < 2)
+        || (unavailableWorkerRatio >= clientShuffleDynamicResourceFactor)) {
+
+        val numPartitions = lifecycleManager.latestPartitionLocation.get(shuffleId).size()
```
Comment (zaynt4606) [inline client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:311]: Is there any problem getting numPartitions(sam with register) from here? 
Is there a better place to get it?

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 4]
```diff
@@ -291,42 +292,71 @@ class ChangePartitionManager(
...
+        val ids = new util.ArrayList((0 until numPartitions).map(Integer.valueOf).asJava)
+        val requestSlotsRes = lifecycleManager.requestMasterRequestSlotsWithRetry(shuffleId, ids)
+
```
Comment (zaynt4606) [inline client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:314]: There are no changes to the requestSlots process in the master. Anything else to notice？

**Comment by zaynt4606** [comment]
cc @FMX @RexXiong

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala (hunk 3 above)
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala]: only requestSlots for the ChangePartitionRequest partitions

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -4845,6 +4846,17 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(false)
 
+  val CLIENT_SHUFFLE_DYNAMIC_RESOURCE_FACTOR: ConfigEntry[Double] =
+    buildConf("celeborn.client.shuffle.dynamicResourceFactor")
+      .categories("client")
+      .version("0.6.0")
+      .doc("The ChangePartitionManager will check whether (unavailable workers / shuffle allocated workers) " +
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4853]: Change the doc to adjust the new implementation.

Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4853]: Done~

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 6]
```diff
@@ -291,42 +292,67 @@ class ChangePartitionManager(
...
+            logInfo(s"ChangePartition requestSlots for $shuffleId failed, have no available slots.")
+          case StatusCode.SUCCESS =>
+            logDebug(
+              s"ChangePartition requestSlots request for workers Success! shuffleId: $shuffleId availableWorkers Info: ${requestSlotsRes.workerResource.keySet()}")
+          case StatusCode.WORKER_EXCLUDED =>
+            logInfo(s"ChangePartition requestSlots request for workers for $shuffleId failed due to all workers be excluded!")
+          case _ => // won't happen
+            throw new UnsupportedOperationException()
+        }
+
```
Comment (zaynt4606) [inline client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:336]: if requestSlotsRes.status is not SUCCESS there might cause NPE...

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 7]
```diff
@@ -291,42 +292,72 @@ class ChangePartitionManager(
...
+    candidates.addAll(snapshotCandidates)
+
+    if (dynamicResourceEnabled) {
+      val shuffleAllocatedWorkers = lifecycleManager.workerSnapshots(shuffleId).size()
+      val unavailableWorkerRatio = 1 - (snapshotCandidates.size * 1.0 / shuffleAllocatedWorkers)
+      if (candidates.size < 1 || (pushReplicateEnabled && candidates.size < 2)
+        || (unavailableWorkerRatio >= dynamicResourceUnavailableFactor)) {
+
+        // get new available workers for the request partition ids
+        val partitionIds = new util.HashSet[Integer]()
```
Comment (RexXiong) [inline client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:313]: ChangePartitionRequest is distinct by partitionId so we can directly add partitionId to  List then.

**Review by RexXiong — CHANGES_REQUESTED** [review]

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala (hunk 7 above)
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala]: Done~

**Review by RexXiong — APPROVED** [review]

**Comment by RexXiong** [comment]
Thanks, merge to main(v0.6.0)

**Pull request closed without merging**